File: scripts/fix_alt_all.mjs
#!/usr/bin/env node

/**
 * リポジトリ内Markdownの画像にaltテキストを付与します（MD045対策）。
 * - 対象: リポジトリ配下の .md（node_modules/.git などは除外）
 * - 処理: コードフェンス外の `![](...)` を `![スクリーンショット](...)` に置換
 *   - `.gif` を含む場合は `![操作手順アニメーション](...)`
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const rootDir = path.resolve(__dirname, '..');

const IGNORE_DIR_NAMES = new Set(['node_modules', '.git', 'dist', '.astro']);

function listMarkdownFilesRecursively(dirPath) {
  /** @type {string[]} */
  const results = [];

  const entries = fs.readdirSync(dirPath, { withFileTypes: true });
  for (const entry of entries) {
    if (entry.name.startsWith('.DS_Store')) continue;

    const fullPath = path.join(dirPath, entry.name);

    if (entry.isDirectory()) {
      if (IGNORE_DIR_NAMES.has(entry.name)) continue;
      results.push(...listMarkdownFilesRecursively(fullPath));
      continue;
    }

    if (entry.isFile() && entry.name.endsWith('.md')) {
      results.push(fullPath);
    }
  }

  return results;
}

function isFenceLine(line) {
  const trimmed = line.trimStart();
  const match = /^(?<fence>`{3,}|~{3,})/.exec(trimmed);
  if (!match) return null;
  return { fence: match.groups.fence, char: match.groups.fence[0], len: match.groups.fence.length };
}

function applyAltFixOutsideFences(markdown) {
  const lines = markdown.split('\n');

  /** @type {{char: string, len: number} | null} */
  let openFence = null;

  const out = lines.map((line) => {
    const fence = isFenceLine(line);
    if (fence) {
      if (!openFence) {
        openFence = { char: fence.char, len: fence.len };
      } else if (openFence.char === fence.char && fence.len >= openFence.len) {
        openFence = null;
      }
      return line;
    }

    if (openFence) return line;

    if (!line.includes('![](')) return line;

    return line.replace(/!\[\]\(([^)]+)\)/g, (_m, inside) => {
      const lowered = String(inside).toLowerCase();
      const alt = lowered.includes('.gif') ? '操作手順アニメーション' : 'スクリーンショット';
      return `![${alt}](${inside})`;
    });
  });

  return out.join('\n');
}

function main() {
  const files = listMarkdownFilesRecursively(rootDir);

  let changedCount = 0;
  let changedImageCount = 0;

  for (const file of files) {
    const original = fs.readFileSync(file, 'utf8');
    if (!original.includes('![](')) continue;

    const updated = applyAltFixOutsideFences(original);
    if (updated !== original) {
      const before = (original.match(/!\[\]\(/g) || []).length;
      const after = (updated.match(/!\[\]\(/g) || []).length;

      fs.writeFileSync(file, updated, 'utf8');
      changedCount++;
      changedImageCount += Math.max(0, before - after);
      console.log(`更新: ${path.relative(rootDir, file)} (images fixed: ${Math.max(0, before - after)})`);
    }
  }

  console.log(`\n✅ altテキストを更新したファイル数: ${changedCount} 件`);
  console.log(`✅ 修正した画像（空alt）数: ${changedImageCount} 件`);
}

main();

File: scripts/python/src/testim_parity/tools/fix_alt_all.py
"""``scripts/tools/fix_alt_all.mjs`` の Python port。

リポジトリ内 markdown の画像に alt テキストを付与する (MD045 対策)。code fence
外の ``![](...)`` だけを対象に、``.gif`` を含む場合は ``操作手順アニメーション``、
それ以外は ``スクリーンショット`` を alt として埋める。
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

from ..project import ROOT_DIR

__all__ = ["apply_alt_fix_outside_fences", "main"]


_IGNORE_DIR_NAMES: frozenset[str] = frozenset({"node_modules", ".git", "dist", ".astro"})
_FENCE_RE = re.compile(r"^(`{3,}|~{3,})")
_EMPTY_ALT_IMAGE_RE = re.compile(r"!\[\]\(([^)]+)\)")


def _list_markdown_files_recursively(dir_path: Path) -> list[Path]:
    """DFS で ``.md`` を走査 (mjs と同順 = dir entries の alphabetic 相当)。"""
    results: list[Path] = []
    try:
        entries = sorted(dir_path.iterdir(), key=lambda p: p.name)
    except OSError, PermissionError:
        return []
    for entry in entries:
        if entry.name.startswith(".DS_Store"):
            continue
        if entry.is_dir():
            if entry.name in _IGNORE_DIR_NAMES:
                continue
            results.extend(_list_markdown_files_recursively(entry))
            continue
        if entry.is_file() and entry.name.endswith(".md"):
            results.append(entry)
    return results


def _fence_info(line: str) -> tuple[str, int] | None:
    """fence 行を検出し ``(char, len)`` を返す (非 fence は None)。"""
    trimmed = line.lstrip()
    match = _FENCE_RE.match(trimmed)
    if not match:
        return None
    fence = match.group(1)
    return (fence[0], len(fence))


def apply_alt_fix_outside_fences(markdown: str) -> str:
    """code fence 外の ``![](...)`` に alt テキストを埋め込む (mjs 等価)。"""
    lines = markdown.split("\n")
    open_fence: tuple[str, int] | None = None
    out: list[str] = []

    for line in lines:
        fence = _fence_info(line)
        if fence is not None:
            if open_fence is None:
                open_fence = fence
            elif open_fence[0] == fence[0] and fence[1] >= open_fence[1]:
                open_fence = None
            out.append(line)
            continue

        if open_fence is not None:
            out.append(line)
            continue

        if "![](" not in line:
            out.append(line)
            continue

        def _replace(match: re.Match[str]) -> str:
            inside = match.group(1)
            alt = "操作手順アニメーション" if ".gif" in inside.lower() else "スクリーンショット"
            return f"![{alt}]({inside})"

        out.append(_EMPTY_ALT_IMAGE_RE.sub(_replace, line))

    return "\n".join(out)


def main(argv: list[str] | None = None) -> int:
    """CLI エントリポイント (exit 0)。mjs と同じく引数を取らない。"""
    _ = argv
    files = _list_markdown_files_recursively(ROOT_DIR)

    changed_count = 0
    changed_image_count = 0
    for file in files:
        try:
            original = file.read_text(encoding="utf-8")
        except OSError:
            continue
        if "![](" not in original:
            continue
        updated = apply_alt_fix_outside_fences(original)
        if updated == original:
            continue
        before = len(_EMPTY_ALT_IMAGE_RE.findall(original))
        after = len(_EMPTY_ALT_IMAGE_RE.findall(updated))
        fixed = max(0, before - after)
        file.write_text(updated, encoding="utf-8")
        changed_count += 1
        changed_image_count += fixed
        try:
            rel = file.relative_to(ROOT_DIR)
        except ValueError:
            rel = file
        print(f"更新: {rel} (images fixed: {fixed})")

    print(f"\n✅ altテキストを更新したファイル数: {changed_count} 件")
    print(f"✅ 修正した画像(空alt)数: {changed_image_count} 件")
    return 0


if __name__ == "__main__":
    sys.exit(main())
